src/gui/touchscreengui.cpp

```diff
@@ -837,7 +870,74 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
 						->getRayFromScreenCoordinates(
 								v2s32(event.TouchInput.X,event.TouchInput.Y));
 			}
-		} else {
+		}
+
+		if (m_joystick_id != -1 && event.TouchInput.ID == m_joystick_id) {
+			double distance = sqrt(
+					(m_pointerpos[event.TouchInput.ID].X - event.TouchInput.X) *
+					(m_pointerpos[event.TouchInput.ID].X - event.TouchInput.X) +
+					(m_pointerpos[event.TouchInput.ID].Y - event.TouchInput.Y) *
+					(m_pointerpos[event.TouchInput.ID].Y - event.TouchInput.Y));
+
+			if ((distance > g_settings->getU16("touchscreen_threshold")) ||
+					(m_joystick_has_really_moved)) {
+				m_joystick_has_really_moved = true;
+
+				// update movement
+				s32 X = event.TouchInput.X;
+				s32 Y = event.TouchInput.Y;
+				s32 dx = X - m_pointerpos[event.TouchInput.ID].X;
+				s32 dy = Y - m_pointerpos[event.TouchInput.ID].Y;
+
+				/* angle in degrees */
+				double angle = acos(dx / distance) * 180 / 3.1415926536;
+				if (dy < 0)
+					angle *= -1;
+				/* rotate to make comparing easier */
+				angle = fmod(angle + 180 + 22.5, 360);
+
+				/* reset state before applying */
+				for (unsigned int i = 0; i < 4; i ++)
+					m_joystick_status[i] = false;
+
+				if (distance <= g_settings->getU16("touchscreen_threshold")) {
+					// do nothing
+				} else if (angle < 45)
```

Comment:
you should invert angle tests i think, because this matches for all others

Reply:
I think after adding `angle` with 180 degree, it should not be negative number.

Reply:
Wait. Let me review my code.